lte/gateway/c/core/oai/tasks/mme_app/CMakeLists.txt

```diff
@@ -70,8 +70,8 @@ target_link_libraries(TASK_MME_APP
     ${CONFIG_LIBRARIES}
     COMMON
     LIB_BSTR LIB_HASHTABLE LIB_DIRECTORYD LIB_SECU LIB_EVENT_CLIENT
-    TASK_NAS TASK_S1AP TASK_SERVICE303 TASK_SCTP_SERVER
-    TASK_S6A TASK_SGS protobuf cpp_redis yaml-cpp redis_utils
+    TASK_NAS TASK_S1AP TASK_SERVICE303 TASK_SCTP_SERVER TASK_S6A TASK_SGS
+    protobuf cpp_redis yaml-cpp redis_utils folly
```

Comment:
@ulaskozat @electronjoe I had to add this link lib here to get my test target to compile. Adding folly to my test's CMakeLists didn't seem to work. Let me know if there's a better approach here.

Reply:
where do you use folly library in your test?

Reply:
I don't quite understand why this is happening, but if I remove folly from this list, I get the following from `make test_oai`:

```
FAILED: test/s1ap_task/s1ap_test 
: && /usr/lib/ccache/c++  -g -DDEBUG_IS_ON=1 -g -O0 -fprofile-arcs -ftest-coverage -DSENTRY_ENABLED=1  -fsanitize=address -fsanitize=undefined -fprofile-arcs -ftest-coverage test/s1ap_task/CMakeFiles/s1ap_test.dir/test_s1ap_handle_new_association.cpp.o  -o test/s1ap_task/s1ap_test  -Wl,-rpath,/usr/local/lib  tasks/s1ap/libTASK_S1AP.a  -lgtest  -lgtest_main  tasks/mme_app/libTASK_MME_APP.a  tasks/nas/libTASK_NAS.a  tasks/sgw/libTASK_SGW.a  tasks/gtpv1-u/libTASK_GTPV1U.a  lib/openflow/controller/libLIB_OPENFLOW_CONTROLLER.a  lib/pipelined_client/libLIB_PIPELINED_CLIENT.a  tasks/sgw_s8/libTASK_SGW_S8.a  tasks/s1ap/libTASK_S1AP.a  tasks/mme_app/libTASK_MME_APP.a  tasks/nas/libTASK_NAS.a  tasks/sgw/libTASK_SGW.a  tasks/gtpv1-u/libTASK_GTPV1U.a  lib/openflow/controller/libLIB_OPENFLOW_CONTROLLER.a  lib/pipelined_client/libLIB_PIPELINED_CLIENT.a  tasks/sgw_s8/libTASK_SGW_S8.a  -lconfig  lib/directoryd/libLIB_DIRECTORYD.a  lib/event_client/libLIB_EVENT_CLIENT.a  /home/vagrant/build/c/magma_common/eventd/libEVENTD.a  tasks/sctp/libTASK_SCTP_SERVER.a  tasks/sgs/libTASK_SGS.a  lib/sgs_client/libLIB_SGS_CLIENT.a  common/redis_utils/libredis_utils.a  lib/secu/libLIB_SECU.a  -lnettle  tasks/s6a/libTASK_S6A.a  lib/s6a_proxy/libLIB_S6A_PROXY.a  -lfluid_base  -lfluid_msg  tasks/service303/libTASK_SERVICE303.a  /home/vagrant/build/c/core/common/service303/libSERVICE303_LIB.a  -lprometheus-cpp  -lgrpc  /usr/lib/x86_64-linux-gnu/libprotobuf.so  -pthread  -lcpp_redis  -ltacopie  -lprotobuf  lib/s8_proxy/libLIB_S8_PROXY.a  lib/mobility_client/libLIB_MOBILITY_CLIENT.a  lib/pcef/libLIB_PCEF.a  lib/mobility_client/libLIB_MOBILITY_CLIENT.a  lib/pcef/libLIB_PCEF.a  common/libCOMMON.a  lib/hashtable/libLIB_HASHTABLE.a  lib/3gpp/libLIB_3GPP.a  lib/itti/libLIB_ITTI.a  tasks/s1ap/libLIB_S1AP.a  tasks/ngap/libLIB_NGAP.a  lib/message_utils/libLIB_MESSAGE_UTILS.a  common/libCOMMON.a  lib/hashtable/libLIB_HASHTABLE.a  lib/3gpp/libLIB_3GPP.a  lib/itti/libLIB_ITTI.a  tasks/s1ap/libLIB_S1AP.a  tasks/ngap/libLIB_NGAP.a  lib/message_utils/libLIB_MESSAGE_UTILS.a  -llfds710  common/glogwrapper/libglogwrapper.a  /usr/local/lib/libsentry.so  -lrt  -lpthread  -lczmq  lib/bstr/libLIB_BSTR.a  /home/vagrant/build/c/magma_common/async_grpc/libASYNC_GRPC.a  /home/vagrant/build/c/core/common/service_registry/libSERVICE_REGISTRY.a  /home/vagrant/build/c/magma_common/config/libMAGMA_CONFIG.a  -lglog  -lgrpc++  -ldl  -lyaml-cpp && :
/usr/bin/ld: tasks/mme_app/libTASK_MME_APP.a(mme_events.cpp.o): in function `report_event(folly::dynamic&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)':
/home/vagrant/magma/lte/gateway/c/core/oai/tasks/mme_app/mme_events.cpp:67: undefined reference to `folly::toJson[abi:cxx11](folly::dynamic const&)'
/usr/bin/ld: tasks/mme_app/libTASK_MME_APP.a(mme_events.cpp.o): in function `attach_success_event':
/home/vagrant/magma/lte/gateway/c/core/oai/tasks/mme_app/mme_events.cpp:78: undefined reference to `folly::dynamic::operator[](folly::dynamic const&) &'
/usr/bin/ld: /home/vagrant/magma/lte/gateway/c/core/oai/tasks/mme_app/mme_events.cpp:78: undefined reference to `folly::dynamic::operator=(folly::dynamic&&)'
/usr/bin/ld: tasks/mme_app/libTASK_MME_APP.a(mme_events.cpp.o): in function `detach_success_event':
/home/vagrant/magma/lte/gateway/c/core/oai/tasks/mme_app/mme_events.cpp:88: undefined reference to `folly::dynamic::operator[](folly::dynamic const&) &'
/usr/bin/ld: /home/vagrant/magma/lte/gateway/c/core/oai/tasks/mme_app/mme_events.cpp:88: undefined reference to `folly::dynamic::operator=(folly::dynamic&&)'
/usr/bin/ld: /home/vagrant/magma/lte/gateway/c/core/oai/tasks/mme_app/mme_events.cpp:89: undefined reference to `folly::dynamic::operator[](folly::dynamic const&) &'
/usr/bin/ld: /home/vagrant/magma/lte/gateway/c/core/oai/tasks/mme_app/mme_events.cpp:89: undefined reference to `folly::dynamic::operator=(folly::dynamic&&)'
/usr/bin/ld: tasks/mme_app/libTASK_MME_APP.a(mme_events.cpp.o): in function `s1_setup_success_event':
/home/vagrant/magma/lte/gateway/c/core/oai/tasks/mme_app/mme_events.cpp:98: undefined reference to `folly::dynamic::operator[](folly::dynamic const&) &'
/usr/bin/ld: /home/vagrant/magma/lte/gateway/c/core/oai/tasks/mme_app/mme_events.cpp:98: undefined reference to `folly::dynamic::operator=(folly::dynamic&&)'
/usr/bin/ld: /home/vagrant/magma/lte/gateway/c/core/oai/tasks/mme_app/mme_events.cpp:100: undefined reference to `folly::dynamic::operator[](folly::dynamic const&) &'
/usr/bin/ld: /home/vagrant/magma/lte/gateway/c/core/oai/tasks/mme_app/mme_events.cpp:100: undefined reference to `folly::dynamic::operator=(folly::dynamic&&)'
/usr/bin/ld: /home/vagrant/magma/lte/gateway/c/core/oai/tasks/mme_app/mme_events.cpp:103: undefined reference to `folly::dynamic::operator[](folly::dynamic const&) &'
/usr/bin/ld: /home/vagrant/magma/lte/gateway/c/core/oai/tasks/mme_app/mme_events.cpp:103: undefined reference to `folly::dynamic::operator=(folly::dynamic&&)'
/usr/bin/ld: tasks/mme_app/libTASK_MME_APP.a(mme_events.cpp.o): in function `folly::dynamic::~dynamic()':
/usr/local/include/folly/dynamic-inl.h:331: undefined reference to `folly::dynamic::destroy()'
collect2: error: ld returned 1 exit status
```

Adding folly to the target_link_libraries list in lte/gateway/c/core/oai/test/s1ap_task/CMakeLists.txt results in the same error (except `-lfolly` is added to the compiler command line)

Reply:
ok, looks like it is being used actually by both S1AP and MME_APP for event reporting. If you add it to CMakeLists.txt for S1AP task, probably it would compile. It is better to include them explicitly to both tasks as both of them are using it.

Reply:
I don't see any references to folly in lte/gateway/c/core/oai/tasks/s1ap; there is a usage of s1_setup_success_event (defined in mme_events.h), but folly is not used in the published interface.

Adding folly as a target_link_libraries in tasks/s1ap/CMakeLists.txt does work, but doesn't feel like it makes as much sense as putting it in tasks/mme_app/CMakeLists.txt where its actually included and used.

Reply:
Okay makes sense.

Reply:
@markjen and @ulaskozat see my notes of pain points trying to add a test myself, including `I was surprised that I had to add linkage to unused libraries in my unit test` -> Linking problems subsection of [this gist](https://gist.github.com/electronjoe/56d6d821b2120166859b4a0d8dcb4409). Seems to be related to Mark's issues here?

Reply:
Worth noting that I tried to fix this linking oddity when I ran into it, by correcting the CMake projects for dependencies that lacked linking - but ran into blockages in CI tests that involve a custom OAI docker environment (more tech debt... that we have so many environments we build and run in and no visibility / experience in so many of them).

Reply:
This was just one more branch on the bonfire of `We need Bazel and static dependencies, dynamic is killing us in our thirty environments`.